cruise-control-core/src/main/java/com/linkedin/cruisecontrol/metricdef/MetricDef.java

```diff
@@ -0,0 +1,126 @@
+  private volatile boolean _doneDefinition = false;
+
+  public MetricDef() {
+    _nextIndex = new AtomicInteger(0);
+    _metricInfoByName = new HashMap<>();
+    _metricInfoByIndex = new ArrayList<>();
+    _metricsToPredict = new HashSet<>();
+  }
+
+  /**
+   * Define the metric.
+   *
+   * @param metricName the name of the metric
+   * @param valueComputingStrategy the {@link AggregationFunction} for this metric.
+   * @return this MetricDef
+   */
+  public synchronized MetricDef define(String metricName, String valueComputingStrategy) {
+    return define(metricName, valueComputingStrategy, false);
+  }
+
+  /**
+   * Define the metric.
+   *
+   * @param metricName the name of the metric
+   * @param valueComputingStrategy the {@link AggregationFunction} for this metric.
+   * @param toPredict whether the metric is a metric to be predicted.
+   * @return this MetricDef
+   */
+  public synchronized MetricDef define(String metricName, String valueComputingStrategy, boolean toPredict) {
+    if (_doneDefinition) {
+      throw new IllegalStateException("Cannot add definition after the metric definition is done.");
+    }
+    CruiseControlUtils.ensureValidString("metricName", metricName);
+    CruiseControlUtils.ensureValidString("valueComputingStrategy", valueComputingStrategy);
+
+    _metricInfoByName.compute(metricName, (k, v) -> {
+      if (v != null) {
+        throw new IllegalArgumentException("Metric " + metricName + " is already defined");
+      }
+      int metricId = _nextIndex.getAndIncrement();
+      if (toPredict) {
+        _metricsToPredict.add(metricId);
+      }
+      return new MetricInfo(metricName, metricId, AggregationFunction.valueOf(valueComputingStrategy.toUpperCase()));
+    });
+    MetricInfo info = _metricInfoByName.get(metricName);
+    _metricInfoByIndex.add(info.id(), info);
+    return this;
+  }
+
+  /**
+   * Finish the metric definition and make the MetricDef immutable.
+   */
+  public synchronized void doneDefinition() {
+    _doneDefinition = true;
+  }
+
+  /**
+   * Get the metric id from the metric name.
+   * @param name the metric name.
+   * @return the {@link MetricInfo} associated with the metric name.
+   */
+  public MetricInfo metricInfo(String name) {
+    MetricInfo info = _metricInfoByName.get(name);
+    if (info == null) {
+      throw new IllegalArgumentException("Metric name " + name + " is not defined. Currently defined metrics are "
+                                             + _metricInfoByName);
+    }
+    return info;
+  }
+
+  /**
+   * @return the {@link MetricInfo} by id;
+   */
+  public MetricInfo metricInfo(int id) {
+    if (id >= _nextIndex.get()) {
+      throw new IllegalArgumentException("Metric Id " + id + " is not defined. Currently defined metrics are "
+                                             + _metricInfoByIndex);
+    }
+    return _metricInfoByIndex.get(id);
+  }
+
+  /**
+   * @return A set of metric ids that are to be predicted.
```

Comment:
Should we clarify in JavaDoc what it means to *predict* a metric? -- i.e. to estimate the metric. It is used to determine what metrics needs to go through the linear regression model and such.

Reply:
This part is not very clear yet and may still subject to change. Let's don't worry to much about this for now.